File: codes/matlab/ch20/weideman_cloot_map.m
function weideman_cloot_map()
%% weideman_cloot_map - Chapter 20: illustrative figure for the
%% Weideman--Cloot (1990) sinh map.
%
% Three panels:
%   (a) the map y = sinh(ell t) for ell in {0.5, 1, 1.5, 2}
%   (b) sinh-mapped grids at fixed N = 64
%   (c) sech(y) shown in y-space and t-space (rescaled) at ell = 1
%
% Author: Dr. Denys Dutykh (Khalifa University, Abu Dhabi, UAE)
% Part of "Computational Etudes: A Spectral Approach"

    script_dir = fileparts(mfilename('fullpath'));
    out_dir = fullfile(script_dir, '..', '..', '..', ...
                       'textbook', 'figures', 'ch20', 'matlab');
    if ~exist(out_dir, 'dir'); mkdir(out_dir); end
    configure_style();
    [NAVY, CORAL, TEAL, ORANGE] = colours();

    ell_values = [0.5, 1.0, 1.5, 2.0];
    palette_a  = {NAVY, CORAL, TEAL, ORANGE};
    N_grid     = 64;
    ell_demo   = 1.0;

    fig = figure('Units', 'inches', 'Position', [1, 1, 13.0, 4.0], 'Color', 'w');
    tl = tiledlayout(1, 3, 'Padding', 'compact', 'TileSpacing', 'compact');

    % Panel (a)
    nexttile(tl); hold on;
    t_dense = linspace(-pi, pi, 401);
    for k = 1:numel(ell_values)
        ell = ell_values(k);
        y_dense = sinh(ell .* t_dense);
        y_max = sinh(ell * pi);
        plot(t_dense, y_dense, 'Color', palette_a{k}, 'LineWidth', 1.4, ...
             'DisplayName', sprintf('$\\ell = %g$, $y_{\\max} \\approx %.1f$', ell, y_max));
    end
    yline(0, '-k', 'LineWidth', 0.4, 'Alpha', 0.5, 'HandleVisibility', 'off');
    xline(-pi, ':', 'Color', [0.5 0.5 0.5], 'LineWidth', 0.6, 'HandleVisibility', 'off');
    xline(+pi, ':', 'Color', [0.5 0.5 0.5], 'LineWidth', 0.6, 'HandleVisibility', 'off');
    xlim([-pi - 0.1, pi + 0.1]); ylim([-30, 30]);
    xlabel('computational coordinate $t \in [-\pi, \pi]$', 'Interpreter', 'latex');
    ylabel('physical coordinate $y$', 'Interpreter', 'latex');
    title('(a) the map $y = \sinh(\ell\, t)$', 'Interpreter', 'latex');
    grid on; box on;
    legend('Location', 'northwest', 'FontSize', 8, 'Interpreter', 'latex');

    % Panel (b)
    nexttile(tl); hold on;
    y_axis = linspace(-30, 30, 601);
    profile = 0.4 ./ cosh(y_axis);
    plot(y_axis, profile + 0.45, 'Color', [0.5 0.5 0.5 0.4], ...
         'LineWidth', 0.8, 'HandleVisibility', 'off');
    text(15, 0.88, 'target sech(y)', 'FontSize', 8, ...
         'Color', [0.5 0.5 0.5]);
    row_y = linspace(0.32, 0.04, numel(ell_values));
    j = 0:N_grid-1;
    t_nodes = -pi + 2 * pi * j / N_grid;
    for k = 1:numel(ell_values)
        ell = ell_values(k);
        y_nodes = sinh(ell .* t_nodes);
        y_show = y_nodes(abs(y_nodes) <= 30);
        plot(y_show, row_y(k) * ones(size(y_show)), '|', ...
             'Color', palette_a{k}, 'MarkerSize', 12, 'LineWidth', 1.1, ...
             'DisplayName', sprintf('$\\ell = %g$', ell));
    end
    xlim([-30, 30]); ylim([0, 1]);
    xlabel('physical coordinate $y$', 'Interpreter', 'latex');
    set(gca, 'YTick', []);
    title(sprintf('(b) sinh-mapped grids at $N = %d$ for several $\\ell$', N_grid), ...
          'Interpreter', 'latex');
    grid on; box on;
    legend('Location', 'southeast', 'FontSize', 9, 'NumColumns', 2, ...
           'Interpreter', 'latex');

    % Panel (c)
    nexttile(tl); hold on;
    y_max = sinh(ell_demo * pi);
    y_phys = linspace(-y_max, y_max, 801);
    sech_in_y = 1.0 ./ cosh(y_phys);
    plot(y_phys, sech_in_y, 'Color', NAVY, 'LineWidth', 1.3, ...
         'DisplayName', '$\mathrm{sech}(y)$ in $y$-space');
    t_axis = linspace(-pi, pi, 801);
    sech_in_t = 1.0 ./ cosh(sinh(ell_demo .* t_axis));
    t_scaled = t_axis * (y_max / pi);
    plot(t_scaled, sech_in_t, '--', 'Color', CORAL, 'LineWidth', 1.3, ...
         'DisplayName', 'same function in $t$-space (scaled)');
    yline(0, '-k', 'LineWidth', 0.4, 'Alpha', 0.5, 'HandleVisibility', 'off');
    xlim([-y_max, y_max]); ylim([-0.1, 1.1]);
    xlabel('physical coordinate $y$ (or scaled $t$)', 'Interpreter', 'latex');
    ylabel('function value', 'Interpreter', 'latex');
    title(sprintf('(c) $\\mathrm{sech}(y)$ in two coordinate frames at $\\ell = %g$', ell_demo), ...
          'Interpreter', 'latex');
    grid on; box on;
    legend('Location', 'northeast', 'FontSize', 9, 'Interpreter', 'latex');

    exportgraphics(fig, fullfile(out_dir, 'weideman_cloot_map.pdf'), ...
                   'ContentType', 'vector');
    exportgraphics(fig, fullfile(out_dir, 'weideman_cloot_map.png'), ...
                   'Resolution', 300);
    close(fig);

    fprintf('[Section 20.x / illustrative figure]  Weideman--Cloot sinh map\n');
    fprintf('  ell values shown: %s\n', mat2str(ell_values));
    fprintf('  y_max = sinh(ell pi) per ell:');
    for k = 1:numel(ell_values)
        fprintf(' %.2f', sinh(ell_values(k) * pi));
    end
    fprintf('\n');
    fprintf('  figure: %s\n', fullfile(out_dir, 'weideman_cloot_map.pdf'));
end

function configure_style()
    set(groot, 'defaultAxesFontName', 'CMU Serif');
    set(groot, 'defaultTextFontName', 'CMU Serif');
    set(groot, 'defaultAxesFontSize', 11);
    set(groot, 'defaultAxesLineWidth', 0.8);
    set(groot, 'defaultLineLineWidth', 1.0);
    set(groot, 'defaultAxesBox', 'on');
end

function [NAVY, CORAL, TEAL, ORANGE] = colours()
    NAVY   = [0x14, 0x2D, 0x6E] / 255;
    CORAL  = [0xE7, 0x4C, 0x3C] / 255;
    TEAL   = [0x16, 0xA0, 0x85] / 255;
    ORANGE = [0xE6, 0x7E, 0x22] / 255;
end

File: codes/julia/ch20/weideman_cloot_map.jl
# weideman_cloot_map.jl
#
# Chapter 20: illustrative figure for the Weideman--Cloot (1990) sinh map.
#
# Three panels:
#   (a) the map y = sinh(ell t) for ell in {0.5, 1, 1.5, 2}
#   (b) sinh-mapped grids at fixed N = 64
#   (c) sech(y) in y-space and t-space (rescaled) at ell = 1
#
# Author: Dr. Denys Dutykh (Khalifa University of Science and Technology)
# Part of "Computational Etudes: A Spectral Approach"

using CairoMakie
using Colors
using FFTW
using Printf

set_theme!(Theme(
    fontsize = 11,
    fonts = (regular = "CMU Serif", bold = "CMU Serif Bold",
             italic  = "CMU Serif Italic"),
    Axis = (xlabelsize = 11, ylabelsize = 11, titlesize = 11,
            xticklabelsize = 10, yticklabelsize = 10,
            spinewidth = 0.8, xtickwidth = 0.8, ytickwidth = 0.8),
    Legend = (labelsize = 9,),
))

const NAVY   = colorant"#142D6E"
const CORAL  = colorant"#E74C3C"
const TEAL   = colorant"#16A085"
const ORANGE = colorant"#E67E22"

const SCRIPT_DIR = @__DIR__
const OUTPUT_DIR = joinpath(SCRIPT_DIR, "..", "..", "..",
                            "textbook", "figures", "ch20", "julia")
mkpath(OUTPUT_DIR)

function make_figure(; ell_values = (0.5, 1.0, 1.5, 2.0),
                     N_grid::Int = 64, ell_demo::Real = 1.0)
    fig = Figure(size = (1100, 800))
    palette_a = [NAVY, CORAL, TEAL, ORANGE]

    ax1 = Axis(fig[1, 1];
        xlabel = "computational coordinate t ∈ [-π, π]",
        ylabel = "physical coordinate y",
        title  = "(a) the map y = sinh(ℓ t)",
        limits = ((-pi - 0.1, pi + 0.1), (-30, 30)))
    t_dense = collect(range(-pi, pi, length = 401))
    for (ell, color) in zip(ell_values, palette_a)
        y_dense = sinh.(ell .* t_dense)
        y_max = sinh(ell * pi)
        lines!(ax1, t_dense, y_dense; color = color, linewidth = 1.4,
               label = "ℓ = $ell, y_max ≈ $(round(y_max, digits=1))")
    end
    hlines!(ax1, [0.0]; color = :black, linewidth = 0.4, alpha = 0.5)
    vlines!(ax1, [-pi, pi]; color = (:gray, 0.6),
            linestyle = :dot, linewidth = 0.6)
    axislegend(ax1, position = :lt, labelsize = 8, framevisible = false)

    ax2 = Axis(fig[1, 2];   # row 1, col 2
        xlabel = "physical coordinate y",
        title  = "(b) sinh-mapped grids at N = $N_grid for several ℓ",
        limits = ((-30, 30), (0, 1)),
        yticksvisible = false, yticklabelsvisible = false)
    y_axis = collect(range(-30, 30, length = 601))
    profile = 0.4 ./ cosh.(y_axis)
    lines!(ax2, y_axis, profile .+ 0.45;
           color = (:gray, 0.4), linewidth = 0.8)
    text!(ax2, 15, 0.88; text = "target sech(y)",
          fontsize = 8, color = (:gray, 0.7))
    row_y = collect(range(0.32, 0.04, length = length(ell_values)))
    j = collect(0:N_grid - 1)
    t_nodes = -pi .+ 2 .* pi .* j ./ N_grid
    for (k, (ell, color)) in enumerate(zip(ell_values, palette_a))
        y_nodes = sinh.(ell .* t_nodes)
        y_show = filter(z -> abs(z) <= 30.0, collect(y_nodes))
        scatter!(ax2, y_show, fill(row_y[k], length(y_show));
                 color = color, marker = :vline,
                 markersize = 14, strokewidth = 0,
                 label = "ℓ = $ell")
    end
    axislegend(ax2, position = :rb, labelsize = 9, framevisible = false,
               nbanks = 2)

    y_max = sinh(ell_demo * pi)
    ax3 = Axis(fig[2, 1];
        xlabel = "physical coordinate y (or scaled t)",
        ylabel = "function value",
        title  = "(c) sech(y) in two coordinate frames at ℓ = $ell_demo",
        limits = ((-y_max, y_max), (-0.1, 1.1)))
    y_phys = collect(range(-y_max, y_max, length = 801))
    sech_in_y = 1.0 ./ cosh.(y_phys)
    lines!(ax3, y_phys, sech_in_y; color = NAVY, linewidth = 1.3,
           label = "sech(y) in y-space")
    t_axis = collect(range(-pi, pi, length = 801))
    sech_in_t = 1.0 ./ cosh.(sinh.(ell_demo .* t_axis))
    t_scaled = t_axis .* (y_max / pi)
    lines!(ax3, t_scaled, sech_in_t; color = CORAL, linewidth = 1.3,
           linestyle = :dash,
           label = "same function in t-space (scaled)")
    hlines!(ax3, [0.0]; color = :black, linewidth = 0.4, alpha = 0.5)
    axislegend(ax3, position = :rt, labelsize = 9, framevisible = false)

    # ---- (d) NEW: Fourier-in-t coefficient diagnostic ----
    N_diag = 96
    j_d = collect(0:N_diag-1)
    t_d = -pi .+ 2 .* pi .* j_d ./ N_diag
    ax4 = Axis(fig[2, 2];
               xlabel = "Fourier mode k",
               ylabel = "|fhat_k| on the t-grid",
               title  = "(d) Fourier-in-t coefficients of sech(y(t)), N = $N_diag",
               yscale = log10,
               limits = (nothing, (1e-17, 1.0)))
    for (ell, color) in zip(ell_values, palette_a)
        fv = 1.0 ./ cosh.(sinh.(ell .* t_d))
        F = abs.(fft(fv)) ./ N_diag
        Fmag = F[1:div(N_diag, 2) + 1]
        ks = collect(0:length(Fmag) - 1)
        scatterlines!(ax4, ks, Fmag .+ 1e-18; color = color,
                      markercolor = :white, strokecolor = color,
                      strokewidth = 1.0, markersize = 3, linewidth = 0.9,
                      label = "ℓ = $ell")
    end
    axislegend(ax4, position = :rt, labelsize = 9, framevisible = false)

    save(joinpath(OUTPUT_DIR, "weideman_cloot_map.pdf"), fig)
    save(joinpath(OUTPUT_DIR, "weideman_cloot_map.png"), fig, px_per_unit = 4)
    return Dict("ell_values" => collect(ell_values),
                "N_grid" => N_grid, "ell_demo" => ell_demo,
                "y_max_per_ell" => [sinh(e * pi) for e in ell_values])
end

if abspath(PROGRAM_FILE) == @__FILE__
    r = make_figure()
    println("[Section 20.x / illustrative figure]  Weideman--Cloot sinh map")
    println("  ell values shown: ", r["ell_values"])
    println("  y_max per ell: ",
            [round(y, digits=2) for y in r["y_max_per_ell"]])
    println("  figure: ", joinpath(OUTPUT_DIR, "weideman_cloot_map.pdf"))
end
